File: android/network/src/test/kotlin/com/ridelink/network/testutil/Vectors.kt
package com.ridelink.network.testutil

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import java.io.File

/**
 * Loads shared golden vectors from `protocol/vectors/` at the repo root.
 *
 * A deliberate twin of `core`'s `com.ridelink.core.testutil.Vectors`: both are five lines of test
 * plumbing, and `core` is a `kotlin("jvm")` module whose test source set an Android library module
 * cannot depend on. Sharing it would mean publishing test fixtures between modules for the sake of
 * a file reader (CLAUDE.md "Shared protocol vectors — not optional" is about the *vector files*
 * being shared, which they are — this is the loader, not the data).
 */
object Vectors {
    private val vectorsDir: File by lazy {
        val path =
            requireNotNull(System.getProperty("ridelink.protocolVectorsDir")) {
                "ridelink.protocolVectorsDir system property not set — see network/build.gradle.kts"
            }
        File(path)
    }

    fun load(relativePath: String): JsonElement {
        val file = File(vectorsDir, relativePath)
        require(file.exists()) { "vector file not found: ${file.absolutePath}" }
        return Json.parseToJsonElement(file.readText())
    }
}

File: ios/Packages/RideLinkPlatform/Tests/RideLinkPlatformTests/TestSupport/Vectors.swift
import Foundation

/// Loads shared golden vectors from `protocol/vectors/` at the repo root.
///
/// A deliberate twin of `RideLinkCoreTests`' loader: both are a few lines of test plumbing, and one
/// package's test target cannot import another's. CLAUDE.md's "Shared protocol vectors — not
/// optional" is about the *vector files* being shared, which they are; this is the reader.
enum Vectors {
    enum VectorsError: Error {
        case repoRootNotFound
    }

    /// Walks upward from this source file's own location (captured at compile time via `#filePath`)
    /// until it finds a directory containing `protocol/vectors`.
    private static func repoRoot(from callerFilePath: String = #filePath) throws -> URL {
        var dir = URL(fileURLWithPath: callerFilePath).deletingLastPathComponent()
        let fileManager = FileManager.default
        while true {
            let candidate = dir.appendingPathComponent("protocol/vectors")
            if fileManager.fileExists(atPath: candidate.path) { return dir }
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path { throw VectorsError.repoRootNotFound }
            dir = parent
        }
    }

    static func loadJSON(_ relativePath: String) throws -> Any {
        let root = try repoRoot()
        let url = root.appendingPathComponent("protocol/vectors").appendingPathComponent(relativePath)
        return try JSONSerialization.jsonObject(with: Data(contentsOf: url), options: [.fragmentsAllowed])
    }
}
